### DISCUSSION TITLE
Naming

### DISCUSSION THREAD

**Discussion by XBagon** [ozankasikci/fantasy-in-rust#26] [General]
I would like to use this thread to discuss the naming of functions, structs, etc. as I think it's kinda important for a library to have memorable and logical names.

My current problem is the name of the parsing functions that use regex. They are called `from_regex`, although they are not converting from "regex" to `Self`, but just from a String with help of regex. I'm unsure if it matters at all to mention that regex is used, I'd even say the methods would be better directly as a `FromStr` implementation.

**Comment by henryksloan** [comment]
I totally agree on the `from_regex` point. Once #20 is merged, we should start applying the same principle to functions named `from_u8` and so on. I agree trait implementations are the way to go, namely `FromStr` and `From<T>`.

I think this thread can also cover the issue of `mod` files being in `src` rather than in the module folders. To my understanding, files like `src/note.rs` should be files like `src/note/mod.rs`.

**Comment by ozankasikci** [comment]
You have a good point with the naming of `from_regex` here. We can refactor this to a more logical name, `FromStr` and `From<T>` makes perfect sense.

However, I'm not in favor of the `mod.rs` naming. It just makes finding the relevant file harder as more and more `mod.rs` files pile up in different directories in the project and it becomes a mess over time. `mod.rs` was actually a restriction from Rust 2015, see this [link](https://doc.rust-lang.org/nightly/edition-guide/rust-2018/module-system/path-clarity.html#no-more-modrs) for the detailed explanation.

**Reply by XBagon** [reply]
I don't know how much less confusing things like `chord::chord::Chord` are. Just stumbled over this and took me some time to realize what was happening. I understand your point for this decision though.

**Reply by ozankasikci** [reply]
Well, it'll never be something like `chord::chord::Chord`. `chord::Chord` is pretty clear IMO.

Also using mod.rs wouldn't change anything here, you'd still need to import the `Chord` struct from a module, so I'm not sure if I follow what you mean.

**Reply by XBagon** [reply]
it is `chord::chord::Chord` internally, that's what the compiler error told me at least 😅. I didn't `pub use` a new thing I put in `src\chord`, which is a module, which has another module `chord` inside of it.

**Reply by ozankasikci** [reply]
Internally it is `chord::chord::Chord` indeed, if you don't `pub use` in `src/chord.rs` file, which is the best practice. We should follow the best practice and import using `chord::Chord` internally as well.

I was actually referring to the API that we expose to the client, and the only way for a client to import the `Chord` struct is `chord::Chord`.
